crates/qasm3/src/build.rs

```diff
@@ -226,42 +226,37 @@ impl BuilderState {
         self.qc.append(py, instruction).map(|_| ())
     }
 
-    fn define_gate(
-        &mut self,
-        _py: Python,
-        ast_symbols: &SymbolTable,
-        decl: &asg::GateDeclaration,
-    ) -> PyResult<()> {
-        let name_id = decl
-            .name()
-            .as_ref()
-            .map_err(|err| QASM3ImporterError::new_err(format!("internal error: {:?}", err)))?;
-        let name_symbol = &ast_symbols[name_id];
-        let pygate = self.pygates.get(name_symbol.name()).ok_or_else(|| {
-            QASM3ImporterError::new_err(format!(
-                "can't handle non-built-in gate: '{}'",
-                name_symbol.name()
-            ))
-        })?;
-        let defined_num_params = decl.params().as_ref().map_or(0, Vec::len);
-        let defined_num_qubits = decl.qubits().len();
-        if pygate.num_params() != defined_num_params {
-            return Err(QASM3ImporterError::new_err(format!(
-                "given constructor for '{}' expects {} parameters, but is defined as taking {}",
-                pygate.name(),
-                pygate.num_params(),
-                defined_num_params,
-            )));
-        }
-        if pygate.num_qubits() != defined_num_qubits {
-            return Err(QASM3ImporterError::new_err(format!(
-                "given constructor for '{}' expects {} qubits, but is defined as taking {}",
-                pygate.name(),
-                pygate.num_qubits(),
-                defined_num_qubits,
-            )));
+    // Map gates in the symbol table to Qiskit gates in the standard library.
+    // Encountering any gates not in the standard library results in raising an exception.
+    // Gates mapped via CustomGates will not raise an exception.
+    fn map_gate_ids(&mut self, _py: Python, ast_symbols: &SymbolTable) -> PyResult<()> {
+        for (name, name_id, defined_num_params, defined_num_qubits) in ast_symbols.gates().iter() {
+            let name = name.to_string();
+            if name == "U" {
+                // The sole built in gate. `gphase` is treated specially.
+                continue;
+            }
```

Comment:
Can we arrange the symbol table to make this check unnecessary?  I'd expect all the built-in symbols to have been populated into the symbol table, so there's no special-case behaviour needed to handle them for a consumer; we could just fetch the symbol and get a gate.

Reply:
Fixed (made the change in my previous comment) in b22b7d7

Reply:
Maybe I misunderstood why this check was here - I'm expecting `U` to be both in `SymbolTable::gates` _and_ to be handled in the same way as other gates in this function.